docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,379 @@
+# Rework of Mesh*Routes matching
+
+* Status: accepted
+
+Technical story: https://github.com/kumahq/kuma/issues/10247
+
+## Context and Problem Statement
+
+At the moment when targeting Mesh*Route you need to put it in topLevel targetRef. 
+We want to iterate over policy design to fully utilize the fact that we are referencing real objects.
+The current implementation has made things a little tricky because policy matching is convoluted where the top level 
+MeshHTTPRoute gets converted to a fake `to`.
+
+## Considered options
+
+- add possibility to put Mesh*Routes in `spec.to[].targetRef`
+- leave Mesh*Routes in top level targetRef as it is now
+
+## Decision Outcome
+
+- add possibility to put Mesh*Routes in `spec.to[].targetRef`
+
+### Positive consequences
+
+- improved policy validation, we can enforce applying only http config on MeshHttpRoute
+- simplifying topLevel targetRef as it will now match only whole proxies
+- no need for specifying topLevel targetRef when targeting `Mesh`
+- easier policy matching code
+
+## Design
+
+The design is grounded in multiple recent MADRs around improvements to policies: 
+
+- [MeshService policy matching](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/046-meshservice-policy-matching.md)
+- [Policy on Namespace](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/052-policy-on-namespace.md) - which introduced consumer/producer policies
+- [Policy matching with real resources](https://github.com/kumahq/kuma/pull/10726/files)
+
+### API
+
+New API for policy that selects Mesh*Route
+
+```yaml
+type: MeshTimeout
+name: timeouts-on-route
+mesh: default
+spec:
+  targetRef:
+    kind: Mesh | MeshSubset # either Mesh, MeshSubset or MeshGateway
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute | MeshTCPRoute # either MeshHTTPRoute or MeshTCPRoute
+        name: route-to-backend
+      default:
+        http:
+          requestTimeout: 5s
+          streamIdleTimeout: 1h
+```
+
+specifying `Mesh` in topLevel targetRef will be optional (this probably will be most used case). Minimal policy, targeting
+`MeshHTTPRoute` will look like this:
+
+```yaml
+type: MeshTimeout
+name: timeouts-on-route
+mesh: default
+spec:
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute
+        name: route-to-backend
+      default:
+        http:
+          requestTimeout: 5s
+          streamIdleTimeout: 1h
+```
+
+To introduce this change we need to deprecate specifying Mesh*Route in topLevel targetRef.
+
+Should we allow selecting MeshGateway in topLevel targetRef? probably this is obsolete as Route will select gateway, and we
+want to apply policy on a route. Is there a need to make it more specific than route? 
```

Comment:
Can we have the same route attached to multiple gateways? Does it make sense to have multiple gateways? 🤔

Reply:
Yes and yes.